urlb44e9843ba_QuasimodoNZ_SWEN222_Game_Project_tgz-pJ8-Game_World_GameWorldLevelBuilderJ8/src/Client_Server/AGServer.java

```diff
@@ -0,0 +1,241 @@
+package Client_Server;
+
+import java.beans.IntrospectionException;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import com.esotericsoftware.kryonet.Connection;
+import com.esotericsoftware.kryonet.Listener;
+import com.esotericsoftware.kryonet.Server;
+import Client_Server.Packets.*;
+import Game_World.GameEngine;
+import Object_Interfaces.GameObject;
+import User_Interface.ServerWindow;
+
+/**
+ * A single server that hosts an Adventure Game.
+ * Handles the incoming connections from all clients, passes
+ * their commands to the game world, then propagates the updated
+ * game world to all client except the originator.
+ *
+ * @author Josh Brake, 300274198, brakejosh
+ */
+public class AGServer {
+
+	private int port;
+	private GameEngine engine;
+	private ServerWindow serverWindow;
+	private Server server;
+	private Map<Connection, String> players;
+
+	public AGServer(int port, final ServerWindow serverWindow, String gameFile) throws IOException {
+
+		if (port < 1 || port > 65535) throw new NumberFormatException("Port out of range"); //port too high/low
+		this.port = port;
+		this.players = new HashMap<Connection, String>();
+		this.serverWindow = serverWindow;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;serverWindow&#39; (index: 1 | type: ServerWindow). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;gameFile&#39;, &#39;port&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).